test/game_test.exs

```diff
@@ -32,21 +32,83 @@ defmodule GameTest do
       assert Game.init("O").player == "O"
     end
 
-    test "cycles through moves until a player wins or ties", context do
+    test "returns game mode attribute" do
+      assert Game.select_mode("1") == :human_v_human
+    end
+
+    test "cycles through moves in human vs. human until a player wins or ties", context do
       Helpers.Stack.setup(["1\n", "2\n", "3\n", "4\n", "5\n", "6\n", "7\n", "8\n", "9\n"])
 
       mock_deps = %{
-        input: MockInput,
-        output: MockOutput,
-        validity: Validity
+        validation: MockValidation,
+        messages: MockMessages,
+        io: MockTTT.IO,
+        player: Player
       }
 
-      game_status = capture_io(fn -> Game.play(context[:initial_game_state], mock_deps) end)
+      game_status =
+        capture_io(fn -> Game.play(context[:initial_game_state], mock_deps, :human_v_human) end)
 
       assert String.contains?(game_status, "game")
 
       Helpers.Stack.teardown()
     end
+
+    test "cycles through moves in human vs. computer until a player wins or ties", context do
+      Helpers.Stack.setup(["1\n", "2\n", "3\n", "4\n", "5\n", "6\n", "7\n", "8\n", "9\n"])
+
+      mock_deps = %{
+        validation: MockValidation,
+        messages: MockMessages,
+        io: MockTTT.IO,
+        player: Player
+      }
+
+      game_status =
+        capture_io(fn -> Game.play(context[:initial_game_state], mock_deps, :human_v_computer) end)
+
+      assert String.contains?(game_status, "game")
+
+      Helpers.Stack.teardown()
+    end
+  end
+
+  describe "when game is one move from ending" do
+    setup do
+      %{
+        one_move_away_state: %State{
+          board: %{
+            {0, 0} => "X",
+            {0, 1} => "O",
+            {0, 2} => "X",
+            {1, 0} => "O",
+            {1, 1} => "X",
+            {1, 2} => "O",
+            {2, 0} => "O",
+            {2, 1} => "X",
+            {2, 2} => 9
+          },
+          player: "X"
+        }
+      }
+    end
+
+    test "finds only available space and enters move", context do
+      assert Game.computer_move(context[:one_move_away_state]) == %State{
+               board: %{
+                 {0, 0} => "X",
+                 {0, 1} => "O",
+                 {0, 2} => "X",
+                 {1, 0} => "O",
+                 {1, 1} => "X",
+                 {1, 2} => "O",
+                 {2, 0} => "O",
+                 {2, 1} => "X",
+                 {2, 2} => "X"
+               },
+               player: "O"
+             }
+    end
```

Comment:
Since the `computer_move` function here just calls through to the `ComputerPlayer.make_move` function, I'd probably move this test into that test file and test `make_move` directly.
Also related to this, as a general guideline I try and keep tests of a given module focused on the functions in that module that other modules depend on ("public" functions). It doesn't always work out, but sometimes if I notice myself writing tests of a private function, it indicates to me that my module might be getting too big and that the private function is showing me a potential seam where I could extract a new module.

Reply:
That's a great point - I've been meaning to ask about what the general process is for TDD'ing as it relates to private functions - is it best to just test the behavior of a public function, and not worry about the details of the private functions it relies on as long as the desired behavior is produced by the public function?

Reply:
I definitely avoid testing private things.  Essentially, if the private function X is being used by Y, and changing X would change Y's output, then any test written for Y could prove that X worked, or provide some confidence, for me, about what X does.

Reply:
Makes sense, I'll start cleaning up the test files/making tests private